teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,565 @@
+
+It's also beneficial that the schemas defined in Chains can follow the field requirements in in-toto attestations. Currently, the type hinting is following a readable formats that users will find it more readable, such as results being populated in [upload-pypi](https://github.com/tektoncd/catalog/blob/f4708d478ee8fac6b5b68347cde087cb7c1d1b1c/task/upload-pypi/0.1/upload-pypi.yaml) and [jib-maven](https://github.com/tektoncd/catalog/blob/6a6f3543fa14d7d840fd13d19ba4452e5e319830/task/jib-maven/0.4/jib-maven.yaml) Catalog Tasks, but these gives difficulties about standarizing signable target schemas, as each signable target's metadata can be faily different from each other. For example, a Git source usually needs a git commit and revision, which are super different from what is needed for OCI images, which need image url and its digest. In-toto attestations provide a way to unify the identifiers for multiple types of signable targets, which is described in the [Use Cases](#use-cases).
+
+Generating the objects that satisfy these schemas, from human readable provenance metadata can be done by writing provenence data using Steps in Tasks into these schema structures in TaskRuns and populate them PipelineRuns' Results, and there is also a discussion around adding a new field `provenance` in PipelineRun and TaskRun, which only allows [trusted Tasks](https://github.com/tektoncd/community/blob/2413ad70a742a6e9103c531e1c5788b2b392a7eb/teps/0091-verified-remote-resources.md#requirements) and Pipelines to generate provenance info in this field. Chains should be able to accomodate what is currently capable in TaskRun and PipelineRun results, and other potential new fields specifically for provenances.
+
+### Notes and Caveats
+
+When these well defined structures grow as richer provenance are pursued, the size of the structure can eventually grow beyond the limit of the container termination message. [TEP 0086](https://github.com/tektoncd/community/pull/521) is addressing this issue.
+
+
+## Design Details
+
+### Signable Targets
+
+All signable targets should be provided in a structure that is well defined in the proposal. If any field is missed, the signable targets will be skipped and an error will be thrown. The users need to follow the naming pattern in order the result objects can be captured.
+
+#### Optional Alternatives for Schemas
+It will be more flexible if users can define the provenance following the defined schemas, but don't necessarily have to follow the naming patterns. Users can provide a result in array type, with name consistent with the signable target, and have a list of result names that are under that category. An example can be like:
+``` yaml
+results:
+        - name: SIGNABLE-VCS
+          value: ["FOO", "BAR"]
```

Comment:
2 alternatives that are probably worth exploring:

1. Consolidate to 2 results - `INPUTS` / `OUTPUTS` (similar to how `IMAGES` is used today)
2. Don't rely on results at all and instead introduce similar fields into the TaskRun status.

I think my preference long-term would be 2 - though we can also hash out what intermediate steps we would need to get there.

Reply:
1. Are you trying to suggest something like this?
```
results:
- name: SIGNABLE_INPUTS
          type: array
```
```
results:
- name: SIGNABLE_OUTPUTS
          type: array
```

Then we have this in TaskRun:
```
results:
- name: SIGNABLE_INPUTS
   value:
   - RESULT_1
   - RESULT_2

- name: SIGNABLE_OUTPUTS
   value:
   - RESULT_3
   - RESULT_4

```
But I think it still matters if we can know the type of the signable target (mave or python), and that would involve something like this with nested objects:
- name: SIGNABLE_INPUTS
   value:
    - name: RESULT_2
       signable_type: VCS
    - name: RESULT_3
       signable_type: IMAGE

Considering nested objects are out of scope for TEP 75, I would probably consider as future support options, but we can also embed this nested object support into option 2.

The current schema has a `signable_type` field that indicates whether the result is an input or an output.

2. I agree that this should be either ultimately put under a TaskRun field or getting this assured that the results are generated by some trusted tasks

Reply:
Discussed offline to refactor the schema. See doc for details